File: app/src/main/java/com/example/pizzaHut/fragments/SearchFragment.java
package com.example.pizzaHut.fragments;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;

import com.example.pizzaHut.adapters.MealAdapter;
import com.example.pizzaHut.adapters.MealAdapter.MealsItemClickListener;
import com.example.pizzaHut.databinding.FragmentSearchBinding;
import com.example.pizzaHut.models.Meal;
import com.example.pizzaHut.responses.MealResponse;
import com.example.pizzaHut.viewmodels.DescriptionViewModel;

import org.jetbrains.annotations.NotNull;

import static android.content.res.Configuration.ORIENTATION_LANDSCAPE;
import static android.content.res.Configuration.ORIENTATION_PORTRAIT;
import static androidx.navigation.Navigation.findNavController;
import static com.example.pizzaHut.fragments.SearchFragmentDirections.fromSearchToDescription;

public class SearchFragment extends Fragment implements MealsItemClickListener {

    private FragmentSearchBinding binding;
    private DescriptionViewModel viewModel;
    private int spanCount;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentSearchBinding.inflate(inflater, container, false);
        spanCount = requireActivity().getResources().getConfiguration().orientation == ORIENTATION_LANDSCAPE ? ORIENTATION_LANDSCAPE : ORIENTATION_PORTRAIT;

        if (!binding.radioBtnMeal.isChecked() && !binding.radioBtnArea.isChecked() && !binding.radioBtnIngredients.isChecked())
            binding.radioBtnMeal.setChecked(true);

        viewModel = new ViewModelProvider(SearchFragment.this).get(DescriptionViewModel.class);
        getMeals();

        return binding.getRoot();
    }

    private void initRecyclerView(MealResponse mealResponse) {
        if (mealResponse != null) {
            MealAdapter mealAdapter = new MealAdapter(mealResponse.getMeals(), SearchFragment.this);
            binding.mealsList.setLayoutManager(new GridLayoutManager(getContext(), spanCount, GridLayoutManager.VERTICAL, false));
            binding.mealsList.setAdapter(mealAdapter);
            binding.mealsList.setHasFixedSize(true);
        }
    }

    private void getMeals() {
        binding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (binding.etSearch.getText() != null) {
                    String search = binding.etSearch.getText().toString();

                    if (binding.radioBtnMeal.isChecked())
                        if (binding.etSearch.getText().length() == 1)
                            viewModel.getMealsByFirstLetter(search.charAt(0)).observe(getViewLifecycleOwner(),
                                    mealResponse -> initRecyclerView(mealResponse));
                        else
                            viewModel.getMealInfo(search).observe(getViewLifecycleOwner(),
                                    mealResponse -> initRecyclerView(mealResponse));

                    else if (binding.radioBtnArea.isChecked())
                        viewModel.getMealsByArea(search).observe(getViewLifecycleOwner(),
                                mealResponse -> initRecyclerView(mealResponse));

                    else if (binding.radioBtnIngredients.isChecked())
                        viewModel.getMealsByIngredient(search).observe(getViewLifecycleOwner(),
                                mealResponse -> initRecyclerView(mealResponse));

                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    @Override
    public void onMealClick(@NotNull Meal meal) {
        findNavController(binding.getRoot()).navigate(fromSearchToDescription(meal.getStrMeal(), null));
    }

    @Override
    public void onMealLongClick(@NotNull String name) {
    }
}

File: app/src/main/java/com/example/foodRecipes/fragments/SearchFragment.kt
package com.example.foodRecipes.fragments

import android.app.Activity
import android.content.res.Configuration.ORIENTATION_LANDSCAPE
import android.content.res.Configuration.ORIENTATION_PORTRAIT
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.Editable
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import androidx.appcompat.widget.AppCompatEditText
import androidx.appcompat.widget.AppCompatTextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView.VERTICAL
import com.example.foodRecipes.R
import com.example.foodRecipes.adapters.MealAdapter
import com.example.foodRecipes.adapters.MealAdapter.MealsItemClickListener
import com.example.foodRecipes.databinding.FragmentMealsBinding
import com.example.foodRecipes.fragments.SearchFragmentDirections.toDescriptionFragment
import com.example.foodRecipes.models.Meal
import com.example.foodRecipes.responses.MealResponse
import com.example.foodRecipes.util.SimpleTextWatcher
import com.example.foodRecipes.viewmodels.SearchViewModel
import java.util.*
import kotlin.collections.ArrayList

class SearchFragment : Fragment(), MealsItemClickListener {

    private val viewModel by viewModels<SearchViewModel>()
    private lateinit var binding: FragmentMealsBinding
    private lateinit var adapter: MealAdapter
    private lateinit var etSearch: AppCompatEditText
    private var meals: List<Meal> = ArrayList()

    private val spanCount: Int
        get() =
            if (resources.configuration.orientation == ORIENTATION_LANDSCAPE)
                ORIENTATION_LANDSCAPE
            else ORIENTATION_PORTRAIT


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentMealsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        activity?.initViews()

        etSearch.addTextChangedListener(object : SimpleTextWatcher() {
            private val handler = Handler(Looper.getMainLooper())
            private lateinit var runnable: Runnable

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (!this::runnable.isInitialized) return

                handler.removeCallbacks(runnable)
            }

            override fun afterTextChanged(s: Editable?) {
                if (s.toString().trim().isNotEmpty()) {
                    handler.postDelayed({
                        if (s?.length == 1) {
                            viewModel.getMealsByFirstLetter(s[0]).observe(viewLifecycleOwner, { response ->
                                initRecyclerView(response)
                            })
                        } else {
                            viewModel.getMealsByName(s.toString()).observe(viewLifecycleOwner, { response ->
                                initRecyclerView(response)
                            })
                        }
                    }, 800)
                }
            }
        })
    }


    private fun Activity.initViews() {
        etSearch = findViewById(R.id.et_search)
        etSearch.visibility = VISIBLE
    }

    private fun initRecyclerView(mealResponse: MealResponse?) {
        adapter = MealAdapter(this@SearchFragment)
        adapter.submitList(mealResponse?.meals)
        binding.mealsList.setHasFixedSize(true)
        binding.mealsList.adapter = adapter
        binding.mealsList.layoutManager = GridLayoutManager(context, spanCount, VERTICAL, false)
        val oldCount = meals.size
        meals = mealResponse!!.meals
        adapter.notifyItemRangeInserted(oldCount, meals.size)
    }

    override fun onMealClick(meal: Meal) = findNavController()
        .navigate(toDescriptionFragment(meal.strMeal, null))

    override fun onMealLongClick(id: String) = Unit
}